tests/integration/cmd_app_test.go

```diff
@@ -27,102 +23,128 @@ var _ = Describe("odo app command tests", func() {
 		helper.CommonAfterEach(commonVar)
 	})
 
-	Context("when running help for app command", func() {
-		It("should display the help", func() {
-			appHelp := helper.Cmd("odo", "app", "-h").ShouldPass().Out()
-			// Trimmed the end of the message string to make it compatible across clusters
-			Expect(appHelp).To(ContainSubstring("Performs application operations related to"))
-		})
+	It("should display the help for app command", func() {
+		appHelp := helper.Cmd("odo", "app", "-h").ShouldPass().Out()
+		// Trimmed the end of the message string to make it compatible across clusters
+		Expect(appHelp).To(ContainSubstring("Performs application operations related to"))
 	})
 
-	Context("when running app delete, describe and list command on fresh cluster", func() {
-		It("should error out display the help", func() {
+	Context("on a fresh new project", func() {
+
+		BeforeEach(func() {
```

Comment:
Why not `JustBeforeEach`?

Reply:
What would be the advantage of using `JustBeforeEach` here?

We have to use `JustBeforeEach` only when really necessary, as the ginkgo doc explains:

```
You can have multiple JustBeforeEaches at different levels of nesting. Ginkgo will first run all the BeforeEaches from the outside in, then it will run the JustBeforeEaches from the outside in. While powerful, this can lead to confusing test suites – so use nested JustBeforeEaches judiciously.

Some parting words: JustBeforeEach is a powerful tool that can be easily abused. Use it well.
```

Reply:
AFAIU spec is the final level in the hierarchy of ginkgo test. For example Describe -> context -> when -> It (Spec). Similarly as part of teardown hierarchy is concerned it will be something like BeforeSuite -> BeforeEach (Describe, context and when level)-> JustBeforeEach (Spec level). The only advantage i can see here is a hierarchical way of applying teardown steps.

Reply:
What you recommend is using `JustBeforeEach` for the last level of hierarchy before `It`, right?

```
Describe("odo app command tests", func() {
	var _ = BeforeEach(func() {
		// #1
	})
	When("creating a new component", func() {
		BeforeEach(func() {
			// #2
		})
		When("odo push is executed", func() {
			JustBeforeEach(func() {
				// #3
			})
			It("...", func() {
				// ...
			})

		})
	})
})
```

1. I cannot see an advantage of using JustBeforeEach instead of BeforeEach, as in both cases, it will be executed in #3.

2. The drawback I can see if if you want to add some tests later, like this, you will be in risk to finish with several `JustBeforeEach` at different levels, which is complex to manage (as explained in the ginkgo doc):

```
Describe("odo app command tests", func() {
	var _ = BeforeEach(func() {
		// #1
	})
	When("creating a new component", func() {
		BeforeEach(func() {
			// #2
		})
		When("odo push is executed", func() {
			JustBeforeEach(func() {
				// #3
			})
			It("...", func() {
				// ...
			})

			When("...", func() {
				JustBeforeEach(func() {
					// #4
				})
				It("...", func() {
					// ...
				})
			})
		})
	})
})
```

Reply:
Thanks for the details. We will document those in out test guidelines 👍